### ISSUE TITLE
dxinterop does not work for my nvidia gpus

### ISSUE THREAD

**Issue by wang-bin** [mpv-player/mpv#2949]
```
[cplayer] Command line options: '-v' '-vo' 'opengl:backend=dxinterop' '-hwdec' 'dxva2' 'E:/media/video/big_buck_bunny_1080p_h264.mov'
[cplayer] mpv git-a1ffd74 (C) 2000-2016 mpv/MPlayer/mplayer2 projects
[cplayer]  built on Mon Mar 14 17:15:58 CST 2016
[cplayer] ffmpeg library versions:
[cplayer]    libavutil       54.31.100
[cplayer]    libavcodec      56.60.100
[cplayer]    libavformat     56.40.101
[cplayer]    libswscale      3.1.101
[cplayer]    libavfilter     5.40.101
[cplayer]    libswresample   1.2.101
[cplayer] ffmpeg version: 2.8.2
[cplayer] 
[cplayer] Configuration: ./waf configure
[cplayer] List of enabled features: asm atomics av-pix-fmt-mmal av-version-info avcodec-chroma-pos-api avframe-metadata avframe-skip-samples build-date c11-tls caca cplayer debug-build direct3d dlopen dos-paths dxva2-hwaccel encoding gl gl-dxinterop gl-win32 glob-win32-replacement iconv jpeg libass libass-osd libav libavdevice libavfilter libbluray libdl libm libswresample lua mingw noexecstack optimize posix-or-mingw resampler sse4-intrinsics stdatomic subprocess tv wasapi win32 win32-internal-pthreads zlib
[global] config path: '' -> 'C:\Users\duncan\AppData\Roaming/mpv'
[global] config path: 'mpv.conf' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/mpv.conf'
[global] config path: 'config' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/config'
[global] config path: 'mpv.conf' -/-> 'C:/bin/mpv.conf'
[global] config path: 'config' -/-> 'C:/bin/config'
[global] config path: 'mpv.conf' -/-> 'C:/bin/mpv/mpv.conf'
[global] config path: 'config' -/-> 'C:/bin/mpv/config'
[cplayer] Setting option 'v' = '' (flags = 8)
[cplayer] Setting option 'vo' = 'opengl:backend=dxinterop' (flags = 8)
[cplayer] Setting option 'hwdec' = 'dxva2' (flags = 8)
[global] config path: 'input.conf' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/input.conf'
[global] config path: 'input.conf' -/-> 'C:/bin/input.conf'
[global] config path: 'input.conf' -/-> 'C:/bin/mpv/input.conf'
[input] Falling back on default (hardcoded) input config
[osc] Loading script @osc.lua...
[global] config path: 'scripts' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/mpv/scripts'
[osc] loading mp.defaults
[osc] loading @osc.lua
[global] config path: 'lua-settings/osc.conf' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/lua-settings/osc.conf'
[global] config path: 'lua-settings/osc.conf' -/-> 'C:/bin/lua-settings/osc.conf'
[global] config path: 'lua-settings/osc.conf' -/-> 'C:/bin/mpv/lua-settings/osc.conf'
[osc] lua-settings/osc.conf not found. 
[cplayer] Run command: define-section, flags=0, args=[showhide, mouse_move script-binding osc/__keybinding1
[cplayer] mouse_leave script-binding osc/__keybinding2
[cplayer] , force]
[cplayer] Run command: enable-section, flags=0, args=[showhide, allow-hide-cursor+allow-vo-dragging]
[cplayer] Run command: define-section, flags=0, args=[input, mouse_btn0 script-binding osc/__keybinding3
[cplayer] shift+mouse_btn0 script-binding osc/__keybinding4
[cplayer] mouse_btn2 script-binding osc/__keybinding5
[cplayer] mouse_btn0_dbl ignore
[cplayer] shift+mouse_btn0_dbl ignore
[cplayer] mouse_btn2_dbl ignore
[cplayer] , force]
[cplayer] Run command: enable-section, flags=0, args=[input, ]
[cplayer] Run command: define-section, flags=0, args=[input_osc, del script-binding osc/__keybinding6
[cplayer] , default]
[cplayer] Run command: enable-section, flags=0, args=[input_osc, allow-hide-cursor+allow-vo-dragging]
[cplayer] Run command: define-section, flags=0, args=[input_forced_osc, , force]
[cplayer] Run command: enable-section, flags=0, args=[input_forced_osc, allow-hide-cursor+allow-vo-dragging]
[cplayer] Run command: disable-section, flags=0, args=[input]
[global] config path: 'fonts' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/fonts'
[global] config path: 'fonts' -> 'C:/bin/fonts'
[osd/libass] Shaper: FriBidi 0.19.6 (SIMPLE) HarfBuzz-ng 1.1.2 (COMPLEX)
[global] config path: 'subfont.ttf' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/subfont.ttf'
[global] config path: 'subfont.ttf' -/-> 'C:/bin/subfont.ttf'
[global] config path: 'subfont.ttf' -/-> 'C:/bin/mpv/subfont.ttf'
[global] config path: 'fonts.conf' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/fonts.conf'
[global] config path: 'fonts.conf' -/-> 'C:/bin/fonts.conf'
[global] config path: 'fonts.conf' -> 'C:/bin/mpv/fonts.conf'
[osd/libass] Setting up fonts...
[osd/libass] Loading font file 'C:/bin/fonts/NotoNaskhArabic-Regular.ttf'
[osd/libass] Loading font file 'C:/bin/fonts/NotoSans-Regular.ttf'
[osd/libass] Loading font file 'C:/bin/fonts/NotoSansHans-Regular.otf'
[osd/libass] Loading font file 'C:/bin/fonts/NotoSansJP-Regular.otf'
[osd/libass] Loading font file 'C:/bin/fonts/NotoSansKR-Regular.otf'
[osd/libass] Loading font file 'C:/bin/fonts/NotoSansSymbols-Regular.ttf'
[osd/libass] Using font provider directwrite
[osd/libass] Done.
[cplayer] Done loading @osc.lua.
[ytdl_hook] Loading script @ytdl_hook.lua...
[global] config path: 'scripts' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/mpv/scripts'
[ytdl_hook] loading mp.defaults
[ytdl_hook] loading @ytdl_hook.lua
[cplayer] Run command: hook-add, flags=0, args=[on_load, 1, 10]
[cplayer] Done loading @ytdl_hook.lua.
[global] config path: 'scripts' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/scripts'
[global] config path: 'scripts' -/-> 'C:/bin/mpv/scripts'
[global] config path: 'watch_later' -> 'C:\Users\duncan\AppData\Roaming/mpv/watch_later'
[cplayer] Playing: E:/media/video/big_buck_bunny_1080p_h264.mov
[cplayer] Running hook: ytdl_hook/on_load
[cplayer] Run command: hook-ack, flags=0, args=[on_load]
[bdmv/bluray] Opening E:/media/video/big_buck_bunny_1080p_h264.mov
[file] Opening E:/media/video/big_buck_bunny_1080p_h264.mov
[file] Stream opened successfully.
[demux] Trying demuxers for level=normal.
[lavf] Found 'mov,mp4,m4a,3gp,3g2,mj2' at score=100 size=2048.
[lavf] avformat_find_stream_info() finished after 460262 bytes.
[demux] Detected file format: mov,mp4,m4a,3gp,3g2,mj2 (libavformat)
[find_files] Loading external files in E:/media/video/
[global] config path: 'sub/' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/sub/'
[global] config path: 'sub/' -/-> 'C:/bin/sub/'
[global] config path: 'sub/' -/-> 'C:/bin/mpv/sub/'
[global] config path: 'audio/' -/-> 'C:\Users\duncan\AppData\Roaming/mpv/audio/'
[global] config path: 'audio/' -/-> 'C:/bin/audio/'
[global] config path: 'audio/' -/-> 'C:/bin/mpv/audio/'
[cplayer]  (+) Video --vid=1 (*) (h264)
[cplayer]  (+) Audio --aid=1 --alang=eng (*) (aac)
[vo/opengl] Setting option 'backend' = 'dxinterop' (flags = 0)
[vo/opengl] Initializing OpenGL backend 'dxinterop'
[vo/opengl] GL_VERSION='4.5.0 NVIDIA 364.51'
[vo/opengl] Detected desktop OpenGL 4.5.
[vo/opengl] GL_VENDOR='NVIDIA Corporation'
[vo/opengl] GL_RENDERER='GeForce GT 750M/PCIe/SSE2'
[vo/opengl] GL_SHADING_LANGUAGE_VERSION='4.50 NVIDIA'
[vo/opengl] Loaded functions for 210/builtin.
[vo/opengl] Loaded functions for 210/builtin.
[vo/opengl] Loaded functions for 300/builtin.
[vo/opengl] Loaded functions for 210/builtin.
[vo/opengl] Loaded functions for 110/GL_EXT_unpack_subimage.
[vo/opengl] Loaded functions for 300/GL_ARB_framebuffer_object.
[vo/opengl] Loaded functions for 300/GL_ARB_vertex_array_object.
[vo/opengl] Loaded functions for 300/GL_ARB_texture_rg.
[vo/opengl] Loaded functions for 320/GL_ARB_sync.
[vo/opengl] Loaded functions for 0/WGL_EXT_swap_control.
[vo/opengl] Loaded functions for 0/WGL_NV_DX_interop.
[vo/opengl] Loaded functions for 430/GL_ARB_debug_output.
[vo/opengl] Loaded functions for 310/GL_ARB_uniform_buffer_object.
[vo/opengl] Filterable half-float textures supported.
[vo/opengl] DX_interop backbuffer size: 1x1
[vo/opengl] DX_interop backbuffer format: 22
[vo/opengl] Couldn't share rendertarget with OpenGL: <Unknown> (0xc007006e)
[vo/opengl/win32] uninit
[cplayer] Error opening/initializing the selected video_out (-vo) device.
[cplayer] Video: no video
[ad] Codec list:
[ad]     lavc:aac - AAC (Advanced Audio Coding)
[ad]     lavc:aac_fixed (aac) - AAC (Advanced Audio Coding)
[ad] Opening audio decoder lavc:aac
[ad] Requesting 1 threads for decoding.
[ad] Selected audio codec: AAC (Advanced Audio Coding) [lavc:aac]
[cplayer] Starting playback...
[cplayer] EOF code: 7  
[ad] Uninit audio decoder.
[cplayer] finished playback, no audio or video data played (reason 4)
[cplayer] 
[cplayer] 
[cplayer] Exiting... (Errors when loading file)
[ytdl_hook] Exiting...
[osc] Exiting...
```

Tested gpu: gt750m, gt540m.

It fails at `DXRegisterObjectNV`.

**Comment by lachs0r** [comment]
I guess https://github.com/mpv-player/mpv/pull/2942 is related.

**Comment by kevmitch** [comment]
Although it appears dxva2 is being used in this case, I don't think #2942  is actually causing this problem since a) that bug sees to only affect AMD, and b) this failure occurs in `context_dxinterop.c` when sharing the D3D surface with opengl via `gl->DXRegisterObjectNV`, rather than opening the D3D device with `gl->DXOpenDeviceNV` in `hwdec_dxva2gldx.c`.

This actually appears to be the same bug as #2620 where the same mysterious HRESULT error code `0xc007006e` is reported (I actually found that bug by searching google for the code).

**Comment by wang-bin** [comment]
It's different with #2942 . I have used dx+gl interop in my player for months, DXOpenDeviceNV is called only once, but `DXRegisterObjectNV` always fail.

**Comment by kevmitch** [comment]
After studying the HRESULT [wikipedia](https://en.wikipedia.org/wiki/HRESULT) and [MSDN](https://msdn.microsoft.com/en-us/library/cc231198.aspx) pages, it appears that `0xc007006e` is `0x8007006e` with the "severity bit" set to "severe" . The latter is translated by [`FormatMessage`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms679351%28v=vs.85%29.aspx) to "The system cannot open the device or file specified." 

Funny enough, `0xc007006e` is not even a legal HRESULT since it sets the severity bit without setting the "NT status value" bit - it should therefore more correctly be `0xd007006e`. Not that any of this helps solve the actual problem. . .

@wang-bin if I'm not mistaken, those are mobile chipsets. Are they part of an optimus setup like in #2620. It might be that optimus somehow does not like dxinterop due to confusion over which device to use as per @rossy's [comment](https://github.com/mpv-player/mpv/issues/2620#issuecomment-166179781). Maybe there are some BIOS or driver settings to completely disable the integrated GPU or otherwise make it look more like a regular discrete GPU? I really have no experience with these things as all my machines are pure Intel.

**Comment by lachs0r** [comment]
Let’s just say this is a duplicate of #2620.
